src/librustc_parse/parser/item.rs

```diff
@@ -1638,26 +1638,32 @@ impl<'a> Parser<'a> {
             .span_to_snippet(self.prev_span)
             .map(|s| s.ends_with(")") || s.ends_with("]"))
             .unwrap_or(false);
-        let right_brace_span = if has_close_delim {
-            // it's safe to peel off one character only when it has the close delim
-            self.prev_span.with_lo(self.prev_span.hi() - BytePos(1))
-        } else {
-            self.prev_span.shrink_to_hi()
-        };
 
-        self.struct_span_err(
+        let mut err = self.struct_span_err(
             self.prev_span,
             "macros that expand to items must be delimited with braces or followed by a semicolon",
-        )
-        .multipart_suggestion(
-            "change the delimiters to curly braces",
-            vec![
-                (self.prev_span.with_hi(self.prev_span.lo() + BytePos(1)), "{".to_string()),
-                (right_brace_span, '}'.to_string()),
-            ],
-            Applicability::MaybeIncorrect,
-        )
-        .span_suggestion(
+        );
+
+        // To avoid ICE, we shouldn't emit actual suggestions when it hasn't closing delims
+        if has_close_delim {
+            err.multipart_suggestion(
+                "change the delimiters to curly braces",
+                vec![
+                    (self.prev_span.with_hi(self.prev_span.lo() + BytePos(1)), '{'.to_string()),
+                    (self.prev_span.with_lo(self.prev_span.hi() - BytePos(1)), '}'.to_string()),
```

Comment:
Feels like there's a missing `Span` method here (`self.prev_span.shift_left(...)`)

Reply:
I couldn't find the `shift_left` method: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.Span.html
Am I still missing something?

Reply:
I'm saying that we could add a higher level operation to avoid dealing with primitives like this. :)

Reply:
Ah, so we can add this tweak as a pub method of `Span`, right?

Reply:
Yeah that would be the idea, if we think it's a good one. cc also @petrochenkov

Reply:
I feel it'd be nice that we'll do it as follow-up work?

Reply:
Works for me.

Reply:
`SourceMap::next_point` let's you advance 1 char at a time (and is now unicode aware). We don't have a method to move left. It is in `SourceMap` instead of `Span` because you need to look at the underlying code for the span to be validated as not falling in the middle of a character.

Reply:
Also, this might work better with `shrink_to_hi` and `shrink_to_lo` instead of `with_*`.